KPIV-52: Get pivot properties from pivot doc id

dhuynh-Kazoo opened the pull request.

jamesaimonetti committed:
KZOO-397: use classifiers as array (#215)

Commit:
PS-137: Doorman update based on recent changes made for customer (#216)

Commit:
KPUS-7: add Authorizing ID and Type into missed_call payload (#217)

icehess committed:
bump orb to use custom couchdb image (#218)

lazedo committed:
KZOO-398 use kapps_call:bridge_diversions/1 for SIP-Diversions

jamesaimonetti committed:
Update types

lazedo committed:
ignore semi attended transfers

* replace/transferee are ignored

icehess committed:
update orb to consider os-release in cache keys (#221)

With upcoming changes, we need another variable to invalidate caches.

This adds /etc/os-release, uname -pmo and others together and uses them
as cache key in CI.

sanjeevie6353 committed:
KCAL-55: Changed message envelope phrase to use on instead of at (#225)

Co-authored-by: sanjeevie <sanjeevie@aio.kazoo.com>

Previously, the message envelope for caller_and_time config would play the prompt as "you have received a call from ::Telephone Number:: at ::DATETIME::".

Now it will pay as "you have received a call from ::Telephone Number:: on ::DATETIME::"

harenson committed:
KCAL-57: Skip recording if auth endpoint is not found (#230)

danielfinke committed:
KCAL-60: fix `cf_flush_dtmf` (#238)

Make sure to persist the cleared DTMF collection on the call using
`cf_exe:set_call/1`

lazedo committed:
KCAL-58 add option to force presence userid

* when using *4 to park a call, the park presence-id is set to *4xxx and is not handled on fast pickup shortcuts
  this commits adds an option enabled by default to override the park presence userid to *3xxx and can
  be opt-out globally or by account-id with systemconfig/callflow key park.force_presence_user

lazedo committed:
handle ignore_progress in route request

jamesaimonetti committed:
Remove the stacktrace macro

lazedo committed:
KZOO-433: revert differentiating behaviour for semi-attended transfers (#247)

looks like freeswitch originating_xfer_wait_thread has fixed known semi-attended transfers and this didn't ensure proper continuation

Commit:
KCAL-62: User recording storage settings are not respected when call retrieving scenario (#251)

Previously, when retrieving a parked call, if recording was enabled by
the device/user/account, kapps_call:start_recording/1 was used without
the recording data. This resulted in not respecting the recording URL for
a device/user/account.

Now, if recording settings are found on the device/user/account, the
data object is passed to kapps_call:start_recording/2 and the
recording is handled as expected for that endpoint.

manushiperera committed:
KCRO-202: Refactor the lists endpoint for one to one support for entry and identity (#254)

List entries were used in dynamic_cid, lookupcidname and cidlistmatch
callflow actions and those modules are updated to retain the existing
functionality on the new data structure.

Commit:
update shipyard config to support rockylinux

Commit:
enable circle-ci support for rocky linux

jamesaimonetti committed:
Splchk of 'transferring' variants

enoskov-ooma committed:
KCAL-66: Include more data in slot config (#258)

Include the number of the caller/callee (depending on call direction)
and the current timestamp when the slot is saved.

jamesaimonetti committed:
KCAL-32: set language on call for voicemail (#263)

Ensures the kapps_call record has the account's language preference
loaded for prompt playback.

When media/support_account_overrides is set to false, the default
language will be chosen for prompts. Calling voicemail expecting the
account's configured language to be used for prompts and hearing the
default language confuses callers.

When loading the mailbox config, ensure the account's language
preference is set on the kapps_call record so prompts will be played
using the expected language.

eezhov-ooma committed:
KZOO-537: Shared VM boxes (#266)

manushiperera committed:
KZOO-549: Use manual presence docs to calculate / update manual presence (#267)

Previously, manual presence records were maintained in a single
document per master account.

Now, the system has been enhanced to track manual presence in separate
documents for each presence ID, stored under the "modb"
database.

Therefore the presence updating and fetching functionalities within
callflow were adjusted to align with this new structure, and presence
document caching has been removed from callflow application.

icehess committed:
KPIV-53: support Pivot serverless processing of a flow (#271)

Callflow Pivot module can now accepts a serverless field and provides
the Pivot flow to be executed.

This removes the dependence of a web server and make using Pivot
serverless possible.

The new `serverless` field is an object with `flow_doc` and
`flow_type` fields.

- `flow_doc`: can be string with a valid PivotML (a twiml extension)
or a valid Callflow JSON

- `flow_type`: indicates the type of specified flow_doc. can be aither
`application/json` for callflow or `application/xml` or `text/xml` for
PivotML.

Using `serverless` or `voice_url` is mutually exclusive, meaning they
can't be used together. In serverless special care must be taken to
use a full URL for any action url throughout the PivotML and relative
URL will not work in serverless.

Flow doc will be serialized to string if it is JSON and then will be
sent to Pivot app to be processed alongside the flow_type.

Pivot will start processing the flow_doc, if any; otherwise it expects
voice_url. The execution will proceed as usual (other than expected
full URL for all action URLs as mentioned above).

jamesaimonetti committed:
KCAL-32: use owner's language preference, if available (#275)

If the voicemail box has an owner, check for the owner's preferred
language; otherwise use the account's langauge for prompts.

Make sure a default is used if account is missing a language

romanat-ooma committed:
KCAL-69: make sure *97 uses account language (#281)

* make sure *97 uses account language

* make sure language is loaded from call

lazedo committed:
fail_on_single_reject is boolean or list of causes

icehess committed:
KPIV-53: adding documentation for basic pivot serverless (#274)

lazedo committed:
KCAL-70 add soft_stop

like stop but ignoring remaining flows

lazedo committed:
KCAL-71 use configuration for exiting cf_user

use soft_stop to ensure cfwd reason is sent back

jamesaimonetti committed:
KCAL-73: don't reverse the key path (#290)

Previously, if the `variable` parameter contained a nested JSON
path (like "foo.bar.baz" or ["foo", "bar", "baz"]) the call to reverse
the list of keys would result in the target doc being searched for a
key path of ["baz", "bar", "foo"].

The fix is to remove the reverse and keep the key path as defined in
the `variable` setting.

pavel-farkas committed:
KCAL-63 Handle call recording for fast-pickup scenario (#289)

* KCAL-63 Handle call recording for fast-pickup scenario

* Corrected doc for cf_park module

---------

Co-authored-by: Pavel Farkas <pfarkas@2600hz.com>

romanat-ooma committed:
import code from media util and ignore sys config val (#294)

icehess commented on src/module/cf_pivot.erl:
```diff
@@ -88,3 +97,13 @@ wait_for_pivot(Data, Call) ->
+        {'ok', Doc} -> Doc;
+        {'error', _R} ->
+            lager:debug("unable to open pivot doc ~s in ~s: ~p", [DocId, AccountDb, _R]),
+            throw({'error', 'not_found'})
```
Don't. log and `cf_exe:continue(Call)`.

icehess commented on src/module/cf_pivot.erl:
```diff
@@ -35,24 +35,33 @@
 -spec handle(kz_json:object(), kapps_call:call()) -> any().
 handle(Data, Call) ->
-    Serverless = kz_json:get_json_value(<<"serverless">>, Data),
+    NewData =
```
I suggest separate this to functions, one is handling the the id and its fetch error if any and the other conitnues as Data.

something like:

```
case kzd_pivot:id(Data) of
    'undefined' ->
        handle_pivot(Data, Call, Data);
    PivotId ->
        case get_pivot_doc(DocId, AccountId) of
            {'ok', JObj} -> 
                handle_pivot(Data, Call, JObj);
            {'error', _} ->
             lolg and cf_congitie
end.
```

jamesaimonetti committed:
KZOO-614: answer call and wait until the call is confirmed answered (#296)

Wait for the CHANNEL_ANSWER event before any `uuid_break` can flush
the queue and the answer is missed (or not even executed on the
switch).

Initially we looked at using kapps_call_command:b_answer_now/1 but
decided against it. Executing with "now" sends the command to the
switch as a sync call instead of normal async. Sync means the command
is executed in the thread of the receiving worker instead of pushing
the command into the session command queue (async). We should avoid
sync if we can.

dhuynh-Kazoo committed:
KWEB-16: Add custom_request_headers option to cf_webhook (#287)

Add custom_http_headers to cf_webhook Hook

* Rename to custom_request_headers

dhuynh-Kazoo committed:
Handler cf_pivot with doc id and normal flow
